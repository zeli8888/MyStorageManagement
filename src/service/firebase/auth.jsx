import {
    GoogleAuthProvider,
    GithubAuthProvider,
    OAuthProvider,
    signInWithPopup,
    setPersistence,
    browserSessionPersistence,
    signInWithEmailAndPassword,
    signOut,
    createUserWithEmailAndPassword,
    sendPasswordResetEmail,
} from 'firebase/auth';
import { firebaseAuth } from './firebaseConfig';

const googleProvider = new GoogleAuthProvider();
const githubProvider = new GithubAuthProvider();
const microsoftProvider = new OAuthProvider('microsoft.com');

microsoftProvider.setCustomParameters({
    prompt: 'select_account',
});

microsoftProvider.addScope('openid');
microsoftProvider.addScope('email');
microsoftProvider.addScope('profile');

// Sign in with Google functionality
export const signInWithGoogle = async () => {
    try {
        await setPersistence(firebaseAuth, browserSessionPersistence);
        const result = await signInWithPopup(firebaseAuth, googleProvider);
        return {
            success: true,
            user: result.user,
            error: null,
        };
    } catch (error) {
        return {
            success: false,
            user: null,
            error: error.message,
        };
    }
};

// Sign in with GitHub functionality
export const signInWithGithub = async () => {
    try {
        await setPersistence(firebaseAuth, browserSessionPersistence);
        const result = await signInWithPopup(firebaseAuth, githubProvider);
        return {
            success: true,
            user: result.user,
            error: null,
        };
    } catch (error) {
        return {
            success: false,
            user: null,
            error: error.message,
        };
    }
};

// Sign in with Microsoft functionality
export const signInWithMicrosoft = async () => {
    try {
        await setPersistence(firebaseAuth, browserSessionPersistence);
        const result = await signInWithPopup(firebaseAuth, microsoftProvider);
        return {
            success: true,
            user: result.user,
            error: null,
        };
    } catch (error) {
        return {
            success: false,
            user: null,
            error: error.message
        };
    }
};

// Sign in with email and password
export async function signInWithCredentials(email, password) {
    try {
        await setPersistence(firebaseAuth, browserSessionPersistence);
        const userCredential = await signInWithEmailAndPassword(
            firebaseAuth,
            email,
            password,
        );
        return {
            success: true,
            user: userCredential.user,
            error: null,
        };
    } catch (error) {
        return {
            success: false,
            user: null,
            error: error.message || 'Failed to sign in with email/password',
        };
    }
}

// Sign out functionality
export const firebaseSignOut = async () => {
    try {
        await signOut(firebaseAuth);
        return { success: true };
    } catch (error) {
        return {
            success: false,
            error: error.message,
        };
    }
};

// Sign up with email and password
export async function signUpWithCredentials(email, password) {
    try {
        const userCredential = await createUserWithEmailAndPassword(firebaseAuth, email, password);
        return { success: true, user: userCredential.user, error: null };
    } catch (error) {
        return { success: false, user: null, error: error instanceof Error ? error.message : "Registration failed" };
    }
}

// Send password reset email
export async function sendPasswordReset(email) {
    try {
        await sendPasswordResetEmail(firebaseAuth, email);
        return { success: true };
    } catch (error) {
        return { success: false, error: error instanceof Error ? error.message : "Password reset failed" };
    }
}

// Auth state observer
export const onAuthStateChanged = (callback) => {
    return firebaseAuth.onAuthStateChanged(callback);
};