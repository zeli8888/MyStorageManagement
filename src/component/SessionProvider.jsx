import { createContext, useState, useEffect } from 'react';
import { onAuthStateChanged } from '../service/firebase/auth';

export const SessionContext = createContext();

const SessionProvider = ({ children }) => {
  const [session, setSession] = useState(null);
  const [loading, setLoading] = useState(true);
  const [rememberMe, setRememberMe] = useState(false);

  useEffect(() => {
    // Returns an `unsubscribe` function to be called during teardown
    const unsubscribe = onAuthStateChanged((user) => {
      if (user) {
        setSession({
          user: {
            name: user.displayName || '',
            email: user.email || '',
            image: user.photoURL || '',
          },
        });
      } else {
        setSession(null);
      }
      setLoading(false);
    });

    return () => unsubscribe();
  }, []);

  return (
    <SessionContext.Provider value={
      {
        session,
        loading,
        setSession,
        setLoading,
        rememberMe,
        setRememberMe
      }}>
      {children}
    </SessionContext.Provider>)
}

export default SessionProvider;
