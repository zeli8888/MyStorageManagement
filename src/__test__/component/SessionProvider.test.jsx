import { render, waitFor, screen, cleanup } from '@testing-library/react'
import { vi, describe, expect, test, beforeEach, afterEach } from 'vitest'
import SessionProvider, { SessionContext } from '../../component/SessionProvider'
import { onAuthStateChanged } from '../../service/firebase/auth'
import '@testing-library/jest-dom/vitest'
import { act } from 'react'

// Mock Firebase auth module
vi.mock('../../service/firebase/auth', () => ({
    onAuthStateChanged: vi.fn()
}))

describe('SessionProvider', () => {
    let mockUnsubscribe
    beforeEach(() => {
        vi.clearAllMocks()
        // Reset mock functions' states
        mockUnsubscribe = vi.fn()
        onAuthStateChanged.mockImplementation((callback) => {
            // Store callback to trigger in tests
            window.__firebaseAuthCallback = callback
            return mockUnsubscribe
        })
    })
    afterEach(cleanup)

    // Base context consumer component
    const TestConsumer = () => {
        return (
            <SessionContext.Consumer>
                {value => (
                    <div>
                        <span data-testid="session">{JSON.stringify(value.session)}</span>
                        <span data-testid="loading">{value.loading.toString()}</span>
                    </div>
                )}
            </SessionContext.Consumer>
        )
    }

    test('Initial loading state', async () => {
        const { asFragment } = render(
            <SessionProvider>
                <TestConsumer />
            </SessionProvider>
        )
        expect(asFragment()).toMatchSnapshot()
        expect(screen.getByTestId('loading')).toHaveTextContent('true')
    })

    test('Successful authentication scenario', async () => {
        render(
            <SessionProvider>
                <TestConsumer />
            </SessionProvider>
        )
        // Mock Firebase auth callback
        const mockUser = {
            displayName: 'John Doe',
            email: 'john@example.com',
            photoURL: 'https://example.com/avatar.jpg'
        }
        act(() => {
            window.__firebaseAuthCallback(mockUser)
        })
        await waitFor(() => {
            const session = JSON.parse(screen.getByTestId('session').textContent)
            expect(session).toEqual({
                user: {
                    name: 'John Doe',
                    email: 'john@example.com',
                    image: 'https://example.com/avatar.jpg'
                }
            })
            expect(screen.getByTestId('loading')).toHaveTextContent('false')
        })
    })

    test('Unauthenticated user scenario', async () => {
        render(
            <SessionProvider>
                <TestConsumer />
            </SessionProvider>
        )
        // Trigger unauthenticated state
        act(() => {
            window.__firebaseAuthCallback(null)
        })
        await waitFor(() => {
            const session = JSON.parse(screen.getByTestId('session').textContent)
            expect(session).toBeNull()
            expect(screen.getByTestId('loading')).toHaveTextContent('false')
        })
    })

    test('Incomplete user information', async () => {
        render(
            <SessionProvider>
                <TestConsumer />
            </SessionProvider>
        )
        // Mock Firebase auth callback
        const mockUser = {
            displayName: null,
            email: null,
            photoURL: null
        }
        act(() => {
            window.__firebaseAuthCallback(mockUser)
        })
        await waitFor(() => {
            const session = JSON.parse(screen.getByTestId('session').textContent)
            expect(session).toEqual({
                user: {
                    name: '',
                    email: '',
                    image: ''
                }
            })
            expect(screen.getByTestId('loading')).toHaveTextContent('false')
        })
    })

    test('Unsubscribe on component unmount', () => {
        const { unmount } = render(
            <SessionProvider>
                <TestConsumer />
            </SessionProvider>
        )
        unmount()
        expect(mockUnsubscribe).toHaveBeenCalledTimes(1)
    })

    test('Update context method', async () => {
        let capturedSetSession
        render(
            <SessionProvider>
                <SessionContext.Consumer>
                    {value => {
                        capturedSetSession = value.setSession;
                        return (
                            <div>
                                <span data-testid="session">{JSON.stringify(value.session)}</span>
                                <span data-testid="loading">{value.loading.toString()}</span>
                            </div>
                        )
                    }}
                </SessionContext.Consumer>
            </SessionProvider>
        )
        // Test if setSession works
        const mockSession = { user: { name: 'Test User2' } }
        act(() => {
            capturedSetSession(mockSession)
        })
        await waitFor(() => {
            const session = JSON.parse(screen.getByTestId('session').textContent)
            expect(session).toEqual(mockSession)
        })
    })
})
