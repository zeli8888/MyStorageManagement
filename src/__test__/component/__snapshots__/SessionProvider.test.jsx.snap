// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SessionProvider > Initial loading state 1`] = `
<DocumentFragment>
  <div>
    <span
      data-testid="session"
    >
      null
    </span>
    <span
      data-testid="loading"
    >
      true
    </span>
  </div>
</DocumentFragment>
`;
